### PR TITLE
Use custom SDPA for decoder-only HF Transformers 

### PR DESCRIPTION
**Pull request by guangy10**
- Register a custom SDPA attention to HF Transformers
- Use the custom SDPA for decoder-only text models _(custom SDPA is optimized for ExecuTorch)_
- Have to bump up transformers version to latest release (`4.51.0`) in order to use the [`AttentionInterface`](https://github.com/huggingface/transformers/blob/main/docs/source/en/attention_interface.md). This has been addressed in https://github.com/huggingface/optimum-executorch/pull/47
- `optimum-cli export executorch` supports custom SDPA

**3x speedup** using custom SDPA for HF smollm2 _(XNNPACK fp32)_:
Screenshot 2025-04-08 at 4 05 10 PM

General applicable to all causal LMs. For encoder-decoder models, it may apply to the self attention layer in the decoder, can make an experiment in a follow-up PR.

### CODE REVIEW

**Comment by guangy10** [comment]
@larryliu0820 FYI, I'm running into the same the issue when using the custom sdpa with eager. The exported model is fine. https://github.com/huggingface/optimum-executorch/actions/runs/14324546808/job/40147624383?pr=46#step:5:137

Checked the shape of the attn output, and they are same. Will dig further by inspecting if output tensors are close tomorrow.

**Comment by guangy10** [comment]
Transformers version bump has been merged in https://github.com/huggingface/optimum-executorch/pull/47

**Comment by guangy10** [comment]
cc: @larryliu0820 @kimishpatel for review

**Comment by larryliu0820** [comment]
Can you change this: https://github.com/huggingface/optimum-executorch/blob/main/optimum/executorch/modeling.py#L181 to use the new Python API: https://pytorch.org/executorch/stable/index.html

**Comment by guangy10** [comment]
> Can you change this: https://github.com/huggingface/optimum-executorch/blob/main/optimum/executorch/modeling.py#L181 to use the new Python API: https://pytorch.org/executorch/stable/index.html

@larryliu0820 Yeah, I'm going to do it in a separate PR.

File: optimum/executorch/attentions/custom_sdpa.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        # assert attention_mask.dim() == 2, f"attention_mask must be a 2D matrix."
+        attention_mask = attention_mask.reshape(-1, max_seq_len)
+        first_row_mask = attention_mask[0, :]
+        # [0, 0, 0, 0, -inf, -inf, -inf, -inf], start_pos = 3
+        start_pos = torch.argmin(first_row_mask).item() - 1
+        output = torch.ops.llama.custom_sdpa(
+            query,
+            key,
+            value,
+            start_pos=start_pos,
```
Comment (kimishpatel) [inline optimum/executorch/attentions/custom_sdpa.py:64]: @larryliu0820 I thought we were gonna make start_pos optional for custom_sdpa. Of course this will have to wait till next release, but we shoudl put a todo here in that case. Advantage of that is that, we dont have to assume whether the sdpa will be used for causal only or not.

Comment (guangy10) [reply optimum/executorch/attentions/custom_sdpa.py:64]: Yeah, I can add a TODO. If this depends on some work in ET first, I'd prefer putting the link to the GitHub Issue

**Comment by guangy10** [comment]
Rebased and fixed conflicts

**Comment by guangy10** [comment]
@larryliu0820 @kimishpatel good to merge?

File: optimum/executorch/attentions/custom_sdpa.py (hunk 1 above)
Comment (kimishpatel) [reply optimum/executorch/attentions/custom_sdpa.py:64]: Ok, lets do a follow up

**Review by kimishpatel — APPROVED** [review]

**Review by michaelbenayoun — APPROVED** [review]

**Comment by guangy10** [comment]
Support export with custom_sdpa using `optimum-cli export executorch`
